### PR TITLE
[TT-10826] self trim oAuth sorted set

### PR DESCRIPTION
**Pull request by jeffy-mathew**
## Description
Add a background job to self trim oAuth sorted set

## Related Issue
https://tyktech.atlassian.net/browse/TT-10826
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [ ] I ensured that the documentation is up to date
- [ ] I explained why this PR updates go.mod in detail with reasoning why it's required
- [ ] I would like a code coverage CI quality gate exception and have explained why

### CODE REVIEW

**Comment by buger** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5907/merge
Commit:  
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/7296754138)

**Comment by buger** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5907/merge
Commit: c6cba5c4477c577b8cb276ff220d583f0856d131 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/7296842550)

**Comment by buger** [comment]
**API tests result - mongo44-murmur64 env: skipped** :no_entry_sign: 
Branch used: refs/pull/5907/merge
Commit: c6cba5c4477c577b8cb276ff220d583f0856d131 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/7296842620)

**Comment by buger** [comment]
**API tests result - postgres15-sha256 env: skipped** :no_entry_sign: 
Branch used: refs/pull/5907/merge
Commit: c6cba5c4477c577b8cb276ff220d583f0856d131 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/7296842620)

**Comment by buger** [comment]
**API tests result - postgres15-murmur64 env: skipped** :no_entry_sign: 
Branch used: refs/pull/5907/merge
Commit: c6cba5c4477c577b8cb276ff220d583f0856d131 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/7296842620)

**Comment by buger** [comment]
**API tests result - mongo44-sha256 env: skipped** :no_entry_sign: 
Branch used: refs/pull/5907/merge
Commit: c6cba5c4477c577b8cb276ff220d583f0856d131 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/7296842620)

**Comment by buger** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5907/merge
Commit: be03d851de4bf3aecdf0f5bdcd658051936bc0cf 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/7297520737)

**Comment by buger** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5907/merge
Commit: be03d851de4bf3aecdf0f5bdcd658051936bc0cf 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/7297520737)

**Comment by buger** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5907/merge
Commit: 9205f101ec5c499f8f47654ebecd6d582bdb3f3e 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/7297703677)

**Comment by buger** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5907/merge
Commit: a594a4e7f891ce1b5dce1c729881768f8941aac6 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/7297865659)

**Comment by buger** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5907/merge
Commit: 8139134291cfdd6b69b0c3c0be5a8d769a71fcf1 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/7298088666)

**Comment by buger** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5907/merge
Commit: 609f5e63fbec644dcce4a7c9715a21a495fb4f67 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/7299088344)

**Comment by buger** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5907/merge
Commit: 01f3e358feb5cd8d673ff4596ce8d3fec8cb36e0 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/7299825709)

**Comment by buger** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5907/merge
Commit: 1446d711f7a89721d669dd3045f18c0d645f55c8 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/7299952037)

**Comment by buger** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5907/merge
Commit: 781b42376247eb69a276c89fc7862c0a463b0bc1 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/7299990055)

**Comment by buger** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5907/merge
Commit: 891822966d558a60d7b0f424fdb92f1252777fec 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/7325380389)

**Comment by buger** [comment]
**API tests result - postgres15-murmur64 env: success** :white_check_mark: 
Branch used: refs/pull/5907/merge
Commit: ce9a876f46d2d6ff879a299df1cd71ce042fcd1c 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/7338153856)

**Comment by buger** [comment]
**API tests result - mongo44-sha256 env: success** :white_check_mark: 
Branch used: refs/pull/5907/merge
Commit: ce9a876f46d2d6ff879a299df1cd71ce042fcd1c 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/7338153856)

**Comment by buger** [comment]
**API tests result - mongo44-murmur64 env: success** :white_check_mark: 
Branch used: refs/pull/5907/merge
Commit: ce9a876f46d2d6ff879a299df1cd71ce042fcd1c 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/7338153856)

**Comment by buger** [comment]
**API tests result - postgres15-sha256 env: success** :white_check_mark: 
Branch used: refs/pull/5907/merge
Commit: ce9a876f46d2d6ff879a299df1cd71ce042fcd1c 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/7338153856)

File: .github/workflows/ci-tests.yml [hunk 1]
```diff
@@ -92,6 +92,11 @@ jobs:
           restore-keys: |
             ${{ runner.os }}-go-
 
+      - name: Run CI benchmarks
+        id: ci-benchmarks
+        run: |
+          ./bin/run-benchmarks.sh 2>&1
```
Comment (titpetric) [inline .github/workflows/ci-tests.yml:98]: This will add up to the CI time. I'd suggest splitting it off to an improvements PR so we create a new job and update the pipeline a bit.

File: bin/run-benchmarks.sh [hunk 2]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+set -e
+
+echo "Running benchmarks"
+
+
+go test -json -benchtime 30s -run='^$' -bench BenchmarkPurgeLapsedOAuthTokens github.com/TykTechnologies/tyk/gateway
```
Comment (titpetric) [inline bin/run-benchmarks.sh:7]:
```suggestion
go test -json -benchtime 30s -run='^$' -bench BenchmarkPurgeLapsedOAuthTokens github.com/TykTechnologies/tyk/gateway

```

File: config/config.go [hunk 3]
```diff
@@ -1066,6 +1066,8 @@ type Config struct {
 
 	// ResourceSync configures mitigation strategy in case sync fails.
 	ResourceSync ResourceSyncConfig `json:"resource_sync"`
+
+	Private Private `json:"-"`
```
Comment (titpetric) [inline config/config.go:1070]: If you use embedding as suggested, you will not have to refactor the references across /gateway, making the change smaller

File: config/private.go [hunk 4]
```diff
@@ -0,0 +1,6 @@
+package config
+
+// Private contains configurations which are private, adding it to be part of config without exposing to customers.
+type Private struct {
+	OAuthTokensPurgeInterval int
```
Comment (titpetric) [inline config/private.go:5]: May as well sort the one other `json:"-"` flag in here, and explicitly add `json:"-"` - we do not want the value to be encodeable into json, so we need to tag fields anyway

File: storage/redis_cluster.go [hunk 5]
```diff
@@ -375,6 +361,23 @@ func (r *RedisCluster) SetRawKey(keyName, session string, timeout int64) error {
 	return nil
 }
 
+func (r *RedisCluster) SetNX(keyName string, value interface{}, timeout int64) (bool, error) {
```
Comment (titpetric) [inline storage/redis_cluster.go:364]: Why are we allowing `interface{}` here?

File: gateway/testutil.go [hunk 6]
```diff
@@ -1009,6 +1009,7 @@ func (s *Test) start(genConf func(globalConf *config.Config)) *Gateway {
 	gw := s.newGateway(genConf)
 	gw.setupPortsWhitelist()
 	gw.startServer()
+
```
Comment (titpetric) [inline gateway/testutil.go:1012]: Unnecessary?

File: gateway/server.go [hunk 7]
```diff
@@ -406,6 +407,17 @@ func (gw *Gateway) setupGlobals() {
 		go gw.flushNetworkAnalytics(gw.ctx)
 	}
 
+	var oAuthTokensPurgeInterval time.Duration
+	if purgeInterval := gw.GetConfig().Private.OAuthTokensPurgeInterval; purgeInterval == 0 {
```
Comment (titpetric) [inline gateway/server.go:411]: FYI even if you embed Private as suggested, it allows using `.Private` like you do here, if you want to be more explicit that's fine.

I would suggest to move this scanner logic onto the type declaration, and provide a getter for the computation. This extends an already long function, so provide this as a `func(*Config)time.Duration` or similar to not extend this function further.

File: gateway/server.go [hunk 8]
```diff
@@ -406,6 +407,17 @@ func (gw *Gateway) setupGlobals() {
...
+		oAuthTokensPurgeInterval = time.Hour
+	} else {
+		oAuthTokensPurgeInterval = time.Second * time.Duration(purgeInterval)
+	}
+
+	oauthTokensPurger := scheduler.NewScheduler("purge-oauth-tokens", oAuthTokensPurgeInterval,
+		log, gw.purgeLapsedOAuthTokens)
+	go oauthTokensPurger.Exec(gw.ctx)
```
Comment (titpetric) [inline gateway/server.go:419]: This may not be the appropriate place. Put it next to pubsubloop or similar goroutine creation. Also, can we use `Start` for the naming

File: gateway/oauth_manager.go [hunk 9]
```diff
@@ -18,13 +18,14 @@ import (
 	"github.com/sirupsen/logrus"
 	"golang.org/x/crypto/bcrypt"
 
+	tykerrors "github.com/TykTechnologies/tyk/internal/errors"
```
Comment (titpetric) [inline gateway/oauth_manager.go:21]: Remove new import block, order tyk imports together

File: internal/scheduler/scheduler.go [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+			s.logger.Error(err)
+			continue
+		}
+
+		s.logger.Infof("execution success")
+	}
+}
+
+func (s *Scheduler) runExecFunc(ctx context.Context) error {
+	tick := time.NewTicker(s.interval)
```
Comment (titpetric) [inline internal/scheduler/scheduler.go:44]: Missed Close

File: storage/redis_cluster.go (hunk 5 above)
Comment (jeffy-mathew) [reply storage/redis_cluster.go:364]: not to restrict the value, it' the same in redis client

File: config/config.go (hunk 3 above)
Comment (jeffy-mathew) [reply config/config.go:1070]: since this is completely internal/private, it's okay to do that refactoring to make it explicit that it's private. right?

File: config/private.go [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+// Private contains configurations which are private, adding it to be part of config without exposing to customers.
+type Private struct {
+	OAuthTokensPurgeInterval int `json:"-"`
+}
+
+func (p Private) GetOAuthTokensPurgeInterval() time.Duration {
+	var purgeInterval time.Duration
+	if p.OAuthTokensPurgeInterval == 0 {
+		purgeInterval = time.Hour
+	} else {
```
Comment (titpetric) [inline config/private.go:14]: Use direct returns and avoid the else to make function smaller

File: gateway/server.go [hunk 12]
```diff
@@ -21,12 +21,13 @@ import (
...
 
-	"github.com/TykTechnologies/tyk/internal/httputil"
-
 	"sync/atomic"
 	textTemplate "text/template"
 	"time"
 
+	"github.com/TykTechnologies/tyk/internal/httputil"
+	"github.com/TykTechnologies/tyk/internal/scheduler"
+
```
Comment (titpetric) [inline gateway/server.go:30]: Unnecessary whitespace here too

File: gateway/server.go [hunk 13]
```diff
@@ -1765,6 +1766,10 @@ func (gw *Gateway) start() {
 		go gw.startPubSubLoop()
 	}
 
+	oauthTokensPurger := scheduler.NewScheduler("purge-oauth-tokens", gw.GetConfig().Private.GetOAuthTokensPurgeInterval(),
+		log, gw.purgeLapsedOAuthTokens)
+	go oauthTokensPurger.Start(gw.ctx)
+
 	conf := gw.GetConfig()
```
Comment (titpetric) [inline gateway/server.go:1773]: There's a conf already declared here, optimize the above gw.GetConfig (use a temp var for it, have NewScheduler on a single line

File: gateway/oauth_manager_test.go [hunk 14]
```diff
@@ -1357,3 +1361,135 @@ func TestPurgeOAuthClientTokensEvent(t *testing.T) {
...
+	storageKey1, storageKey2 := fmt.Sprintf("%s%s", prefixClientTokens, clientID1),
+		fmt.Sprintf("%s%s", prefixClientTokens, clientID2)
+
+	storageManager := ts.Gw.getGlobalMDCBStorageHandler(generateOAuthPrefix(spec.APIID), false)
+	storageManager.Connect()
+
+	assertTokensLen(t, storageManager, storageKey1, 3)
+	assertTokensLen(t, storageManager, storageKey2, 3)
+
+	time.Sleep(time.Second * 3)
```
Comment (titpetric) [inline gateway/oauth_manager_test.go:1391]: This is still a read flag, we shouldn't sleep this much

File: internal/scheduler/scheduler_test.go [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+	interval := 1 * time.Second
+
+	s := scheduler.NewScheduler("testScheduler", interval, logger, execFunc)
+
+	assert.NotEmpty(t, s)
+}
+
+func TestExec(t *testing.T) {
+	t.Run("success", func(t *testing.T) {
+		t.Skip()
```
Comment (titpetric) [inline internal/scheduler/scheduler_test.go:28]: test.Flaky, the skip there is conditional on CI=true/false.

File: config/private.go [hunk 16]
```diff
@@ -0,0 +1,16 @@
+package config
+
+import "time"
+
+// Private contains configurations which are private, adding it to be part of config without exposing to customers.
+type Private struct {
+	OAuthTokensPurgeInterval int `json:"-"`
```
Comment (titpetric) [inline config/private.go:7]: The field will need a comment per our doc conventions.

I'd like to see Config.OriginalPath moved here as it's the single field in Config that belongs here.

File: storage/redis_cluster.go (hunk 5 above)
Comment (titpetric) [reply storage/redis_cluster.go:364]: The redis client is generic, but we are not, you don't need it to be an interface? As it's intended purpose is a shared-lock mechanism, could we just name it `Lock(key string, timeout int64) (bool, error)`?

File: gateway/server.go (hunk 13 above)
Comment (jeffy-mathew) [reply gateway/server.go:1773]: Ideally I would make it to use across the whole func

File: gateway/oauth_manager_test.go (hunk 14 above)
Comment (jeffy-mathew) [reply gateway/oauth_manager_test.go:1391]: I understand the significant delay introduced here, but we could set `oauth_token_expire` and `oauth_token_expire_retain_period` to be 1, which can reduce it to 2 seconds and not less than that, since they're values in seconds

**Review by titpetric — APPROVED** [review]
